projects/bootloader/scripts/can_datagram.py

```diff
@@ -0,0 +1,294 @@
+        self._check_kwargs(**kwargs)
+
+        self._protocol_version = PROT_VER & 0xff
+        self._datagram_type_id = kwargs["datagram_type_id"] & 0xff
+
+        self._node_ids = []
+        for val in kwargs["node_ids"]:
+            self._node_ids.append(val & 0xff)
+
+        self._data = kwargs["data"]
+
+    @classmethod
+    def deserialize(cls, datagram_bytearray):
+        """This function returns an instance of the class from a bytearray."""
+        assert isinstance(datagram_bytearray, bytearray)
+
+        # "theoretical" lower limit:
+        # 1 (prot) + 4 (crc32) + 1 (type) + 1 (num nodes) + 0 (nodes) + 2 (data size) + 0 (data)
+        #   = 9
+        if len(datagram_bytearray) < MIN_BYTEARRAY_SIZE:
+            raise DatagramTypeError(
+                "Invalid Datagram format from bytearray: Does not meet minimum size requirement")
+
+        protocol_version = datagram_bytearray[PROTOCOL_VERSION_OFFSET]
+        crc32 = datagram_bytearray[CRC_32_OFFSET:DATAGRAM_TYPE_OFFSET]
+        datagram_type_id = datagram_bytearray[DATAGRAM_TYPE_OFFSET]
+
+        num_node_ids = datagram_bytearray[NUM_NODE_ID_OFFSET]
+
+        if len(datagram_bytearray) < MIN_BYTEARRAY_SIZE + num_node_ids:
+            raise DatagramTypeError("Invalid Datagram format from bytearray: Not enough node ids")
+
+        node_ids = list(datagram_bytearray[NODE_ID_OFFSET:NODE_ID_OFFSET + num_node_ids])
+
+        data_size = cls._convert_from_bytearray(cls,
+                                                datagram_bytearray[NODE_ID_OFFSET + num_node_ids:], 2)
+
+        if len(datagram_bytearray) != MIN_BYTEARRAY_SIZE + num_node_ids + data_size:
+            raise DatagramTypeError("Invalid Datagram format from bytearray: Not enough data bytes")
+
+        data = datagram_bytearray[NODE_ID_OFFSET + num_node_ids + DATA_SIZE_SIZE:]
+
+        exp_crc32 = cls._calculate_crc32(cls, datagram_type_id, node_ids, data)
+
+        if not exp_crc32 == crc32:
+            raise DatagramTypeError("Invalid Datagram format from bytearray: Invalid crc32")
+
+        return cls(protocol_version=protocol_version,
+                   datagram_type_id=datagram_type_id, node_ids=node_ids, data=data)
+
+    def serialize(self):
+        """This function returns a bytearray based on set data."""
+
+        node_crc32 = zlib.crc32(bytearray(self._node_ids))
+        node_crc32 = self._convert_to_bytearray(node_crc32, 4)
+        data_crc32 = zlib.crc32(self._data)
+        data_crc32 = self._convert_to_bytearray(data_crc32, 4)
+
+        crc32_array = bytearray([self._datagram_type_id,
+                                 len(self._node_ids),
+                                 *node_crc32,
+                                 len(self._data) & 0xff,
+                                 (len(self._data) >> 8) & 0xff,
+                                 *data_crc32])
+        # Update the crc32
+        crc32 = zlib.crc32(crc32_array)
+        crc32 = self._convert_to_bytearray(crc32, 4)
```

Comment:
Note to self: I forgot to move this crc32 logic to the function I made for it... I will do that.

Reply:
Do this!